### PR TITLE
ChangeSignature: select parameter based on cursor position

### PR DESCRIPTION
**Pull request by jcouv**
Previously, the ChangeSignature dialog would always open with the default selection (ie. the first movable or removable parameter).
With this PR, if you select a parameter in a declaration and invoke ChangeSignature, that parameter will be selected.

image

image

If this PR is ok so far, I'll make the same change on the VB side.

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs [hunk 1]
```diff
@@ -265,7 +265,8 @@ public static void M(this C goo, int y, int x, string c = ""test_c"", string b =
     { }
 }";
 
-            await TestChangeSignatureViaCommandAsync(LanguageNames.CSharp, markup, updatedSignature: permutation, expectedUpdatedInvocationDocumentCode: updatedCode);
+            await TestChangeSignatureViaCommandAsync(LanguageNames.CSharp, markup, updatedSignature: permutation,
+                expectedUpdatedInvocationDocumentCode: updatedCode, expectedSelectedIndex: 0);
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/ChangeSignature/ReorderParametersTests.cs]: :memo: although the parameter config has `0` for the `SelectedIndex`, the UI dialog will make an adjustment and select parameter `y` instead before the `this` parameter cannot be moved or removed.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 2]
```diff
@@ -122,13 +123,60 @@ internal sealed class CSharpChangeSignatureService : AbstractChangeSignatureServ
...
+        private static int TryGetSelectedIndexFromDeclaration(int position, SyntaxNode matchingNode)
+        {
+            var parameters = matchingNode.ChildNodes().OfType<BaseParameterListSyntax>().SingleOrDefault();
+            var selectedIndex = 0;
+            if (parameters != null)
+            {
+                selectedIndex = GetParameterIndex(parameters.Parameters, position);
+            }
+
+            return selectedIndex;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: nit:

```return parameters == null ? 0 : ...```

File: src/EditorFeatures/TestUtilities/ChangeSignature/AbstractChangeSignatureTests.cs [hunk 3]
```diff
@@ -62,7 +62,8 @@ protected override ParseOptions GetScriptOptions()
             string expectedErrorText = null,
             int? totalParameters = null,
             bool verifyNoDiagnostics = false,
-            ParseOptions parseOptions = null)
+            ParseOptions parseOptions = null,
+            int expectedSelectedIndex = -1)
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/ChangeSignature/AbstractChangeSignatureTests.cs]: 💡 Might be better to default this to `0` so it's always covered by testing.

File: src/EditorFeatures/TestUtilities/ChangeSignature/ChangeSignatureTestState.cs [hunk 4]
```diff
@@ -80,6 +81,15 @@ public ChangeSignatureResult ChangeSignature()
                 CancellationToken.None);
         }
 
+        public ParameterConfiguration GetParameterConfiguration()
+        {
+            WpfTestRunner.RequireWpfFact($"{nameof(AbstractChangeSignatureService.ChangeSignature)} currently needs to run on a WPF Fact because it's factored in a way that tries popping up UI in some cases.");
+
+            var context = ChangeSignatureService.GetContextAsync(InvocationDocument, _testDocument.CursorPosition.Value, restrictToDeclarations: false, CancellationToken.None)
+                .WaitAndGetResult(CancellationToken.None);
```
Comment (sharwell) [inline src/EditorFeatures/TestUtilities/ChangeSignature/ChangeSignatureTestState.cs]: 💡 Would be better to make the containing method asynchronous.

File: src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs [hunk 5]
```diff
@@ -122,13 +123,19 @@ internal sealed class CSharpChangeSignatureService : AbstractChangeSignatureServ
...
 
             var symbolInfo = semanticModel.GetSymbolInfo(matchingNode, cancellationToken);
-            return symbolInfo.Symbol ?? symbolInfo.CandidateSymbols.FirstOrDefault();
+            return (symbolInfo.Symbol ?? symbolInfo.CandidateSymbols.FirstOrDefault(), 0);
+        }
+
+        private static int TryGetSelectedIndexFromDeclaration(int position, SyntaxNode matchingNode)
+        {
+            var parameters = matchingNode.ChildNodes().OfType<BaseParameterListSyntax>().SingleOrDefault();
+            return parameters == null ? 0 : GetParameterIndex(parameters.Parameters, position);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs]: 💡 Consider inverting this conditional to emphasize the success case.

File: src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs [hunk 6]
```diff
@@ -533,5 +533,40 @@ private static SignatureChange CreateCompensatingSignatureChange(ISymbol declara
 
             return parametersToPermute;
         }
+
+        /// <summary>
+        /// Given the cursor position, find which parameter is selected.
+        /// Returns 0 as the default value. Note that the ChangeSignature dialog adjusts the selection for
+        /// the `this` parameter in extension methods (the selected index won't remain 0).
+        /// </summary>
+        protected static int GetParameterIndex<TNode>(SeparatedSyntaxList<TNode> parameters, int position) where TNode : SyntaxNode
```
Comment (sharwell) [inline src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs]: :bulb:

```suggestion
        protected static int GetParameterIndex<TNode>(SeparatedSyntaxList<TNode> parameters, int position)
            where TNode : SyntaxNode
```

File: src/EditorFeatures/TestUtilities/ChangeSignature/AbstractChangeSignatureTests.cs (hunk 3 above)
Comment (jcouv) [reply src/EditorFeatures/TestUtilities/ChangeSignature/AbstractChangeSignatureTests.cs]: I tried, but in a number of tests the `ParameterConfiguration` is null. 
I don't want to make the assertion conditional on null, as that could hide issues, so I think the current design is safer.

---
In reply to: [260739080](https://github.com/dotnet/roslyn/pull/33453#discussion_r260739080) [](ancestors = 260739080)

**Comment by jcouv** [comment]
@sharwell Thanks for the review.
FYI, I addressed all of your suggestions except for one. Left a comment to explain why that one didn't work well.
I'll go ahead and merge once CI is green.

**Pull request merged**